Mathlib/LinearAlgebra/Matrix/SVD/Defs.lean

```diff
@@ -0,0 +1,479 @@
+
+namespace Matrix
+
+/-- The right eigenvectors of a matrix A corresponding to its non-zero eigenvaules -/
+noncomputable def svdV₁ (A : Matrix (Fin M) (Fin N) 𝕂) : Matrix (Fin N) (Fin (A.rank)) 𝕂 :=
+  ((reindex (Equiv.refl (Fin N)) (eigenColumnEquiv A))
+    (isHermitian_transpose_mul_self A).eigenvectorMatrix).toColumns₁
+
+/-- The right eigenvectors of a matrix A corresponding to the zero eigenvaules of the matrix AᴴA -/
+noncomputable def svdV₂ (A : Matrix (Fin M) (Fin N) 𝕂) : Matrix (Fin N) (Fin (N - A.rank)) 𝕂 :=
+  ((reindex (Equiv.refl (Fin N)) (eigenColumnEquiv A))
+    (isHermitian_transpose_mul_self A).eigenvectorMatrix).toColumns₂
+
+/-- The diagonal matrix containing the non-zero eigenvalues of matrix Aᴴ * A These are also the
+squares of the non-zero singular values of the matrix A. Note that these are the same elements as in
+`svdμ'` but permuted in some arbitrary order -/
+noncomputable def svdμ (A : Matrix (Fin M) (Fin N) 𝕂) : Matrix (Fin A.rank) (Fin A.rank) ℝ :=
+  (reindex
+    (finRankEquivEigsConjTransposeMulSelf A).symm
+    (finRankEquivEigsConjTransposeMulSelf A).symm)
+  (diagonal (fun (i : {a // (isHermitian_transpose_mul_self A).eigenvalues a ≠ 0}) =>
+      (isHermitian_transpose_mul_self A).eigenvalues i))
+
+/-- The diagonal matrix containing the non-zero eigenvalues of matrix A * Aᴴ These are also the
+squares of the non-zero singular values of the matrix A. Note that these are the same elements as in
+`svdμ` but permuted in some arbitrary order -/
+noncomputable def svdμ' (A : Matrix (Fin M) (Fin N) 𝕂) : Matrix (Fin A.rank) (Fin A.rank) ℝ :=
+  (reindex
+    (finRankEquivEigsMulConjTranspose A).symm
+    (finRankEquivEigsMulConjTranspose A).symm)
+  (diagonal (fun (i : {a // (isHermitian_mul_conjTranspose_self A).eigenvalues a ≠ 0}) =>
+      (isHermitian_mul_conjTranspose_self A).eigenvalues i))
+
+/-- The diagonal matrix containing the non-zero singular values of matrix A. These are also the
+square roots of the non-zero eigenvalues of the matrix Aᴴ * A. -/
+noncomputable def svdσ (A : Matrix (Fin M) (Fin N) 𝕂) : Matrix (Fin A.rank) (Fin A.rank) ℝ :=
+  (reindex
+    (finRankEquivEigsConjTransposeMulSelf A).symm
+    (finRankEquivEigsConjTransposeMulSelf A).symm)
+  (diagonal (fun (i : {a // (isHermitian_transpose_mul_self A).eigenvalues a ≠ 0}) =>
+      Real.sqrt ((isHermitian_transpose_mul_self A).eigenvalues i)))
+
+/-- The left eigenvectors of a matrix A corresponding to its non-zero eigenvaules, obtained as the
+image of the corresponding right eigenvectors. The transformation is given by the matrix itself and
+scaling by the singular values. -/
+noncomputable def svdU₁ (A : Matrix (Fin M) (Fin N) 𝕂) : Matrix (Fin M) (Fin A.rank) 𝕂 :=
+  A  *  A.svdV₁  *  (A.svdσ.map (algebraMap ℝ 𝕂))⁻¹
+
+/-- The left eigenvectors of a matrix A corresponding to its non-zero eigenvaules, obtained directly
+from the eigendecomposition of the AAᴴ matrix. These do NOT share the same ordering as `svdU₁`. -/
```

Comment:
U1' (with dash) is kind of a duplicate but it makes proving the `reduced_spectral_theroem'` (with dash) much easier. 
There are two differences between U1 and U1':

1. They do not share the same order because of how mathlib defines eigenvalues (some arbitrary order). Your proposed reordering is good for this case.
2. With repeated/duplicate eigenvalues each eigenvalue has an eigenspace with Dim > 1 and the eigenvectors can be basis of that eigenspace. Hence the contents of the eigenvector matrices need not be the same, even after disregarding permutations. Hence no reordering will clean this second issue.
 
Hence why I chose to have two definitions.